### ISSUE TITLE
A way to execute SQL scripts before / after "main" deployment process

### ISSUE THREAD

**Issue by Halky** [Rhetos/Rhetos#149]
**Is your feature request related to a problem? Please describe.**
There is no simple and clean way to configure SQL / database features with Rhetos. Example would be enabling FileStream feature or changing Snapshot isolation level.
Additionally it is not possible to test database with standard deployment process once "main" deployment is done.

**Describe the solution you'd like**
A way to define set of (SQL) script with options to execute them either before or after the (main) deployment process. Each script would have options to write warning via DeployPackages (yellow text) or throw exception (red text with aborting deployment process) or to reconfigure database.
Structure of the script would be
```
Setup (should I be executed before or after the "main" deployment process)
Test (SQL commands)
Action if test failed (writing warning / throwing exception / executing additional commands)
Action if test passed (writing white text to Deploy packages)
```

**Comment by bantolov** [comment]
Part of these requirements could be solved with **AfterDeploy scripts** - it's an alternative to the DataMigration scripts, but better suited for infrastructure setup and some data initialization. This feature is provided by a Rhetos plugin package "Rhetos.AfterDeploy". In the [Readme.md file](https://github.com/Rhetos/AfterDeploy/blob/master/Readme.md) there is a description of differences between AfterDeploy and DataMigration scripts.

A more flexible approach is to write C# code that executes the testing scripts:
Rhetos allows development of custom plugins that are executed on each deployment - IServerInitializer. You can use it to develop the testing suite as described in the question.
Here is an example of such plugin: [ExecuteSqlAfterDeploy](https://github.com/Rhetos/AfterDeploy/blob/master/Source/Rhetos.AfterDeploy/ExecuteSqlAfterDeploy.cs). It is **just one class** that implements interface `Rhetos.Extensibility.IServerInitializer`. It's executed at the end of each deployment, where you could read scripts from some folder, execute them on database and check the results.

To execute infrastructeu rure scripts *before* the database upgrade, you need to implement `Rhetos.Extensibility.IGenerator` interface. It is intended for plugins that build the application, but can also be used to execute SQL code on database.

**Comment by bantolov** [comment]
Currently there is no simple plugin system for executing SQL scripts *before* the database upgrade (aside from datamigration scripts), but it will be available soon: The contribution in pull request #135 will allow this kind of plugins in a similar way as described above, by implementing a single interface.

**Comment by bantolov** [comment]
With Rhetos v2.12, there is a new type of plugins available that might also help with this issue:

A plugin class that implements `IConceptDataMigration` can provide SQL scripts to be executed before or after data-migration scripts. For example, this plugin will insert a record in Common.Log at the beginning of each deployment:

```cs
[Export(typeof(IConceptDataMigration))]
public class ExecuteSomethingBeforeDataMigration : IConceptDataMigration<InitializationConcept>
{
    public void GenerateCode(InitializationConcept concept, IDataMigrationScriptBuilder codeBuilder)
    {
        string sql = @"IF OBJECT_ID('Common.Log') IS NOT NULL
            INSERT INTO Common.Log (Action) SELECT 'Something executed';";

        codeBuilder.AddBeforeDataMigrationScript(sql);
    }
}
```

This plugin could be modified to load SQL scripts from the NuGet packages that are included in the Rhetos application. and execute all scripts from certain subfolders.
